google-cloud-clients/google-cloud-contrib/google-cloud-spanner-jdbc/src/main/java/com/google/cloud/spanner/jdbc/ReadWriteTransaction.java

```diff
@@ -0,0 +1,777 @@
+    if (retryAbortsInternally) {
+      addRetryStatement(new FailedUpdate(this, e, update));
+    }
+  }
+
+  /** Registers the statements as a batch of updates that should return an error during a retry. */
+  private void createAndAddFailedBatchUpdate(SpannerException e, Iterable<Statement> updates) {
+    if (retryAbortsInternally) {
+      addRetryStatement(new FailedBatchUpdate(this, e, updates));
+    }
+  }
+
+  /**
+   * Adds a statement to the list of statements that should be retried if this transaction aborts.
+   */
+  private void addRetryStatement(RetriableStatement statement) {
+    Preconditions.checkState(
+        retryAbortsInternally, "retryAbortsInternally is not enabled for this transaction");
+    statements.add(statement);
+  }
+
+  /**
+   * Handles an aborted exception by checking whether the transaction may be retried internally, and
+   * if so, does the retry. If retry is not allowed, or if the retry fails, the method will throw an
+   * {@link AbortedException}.
+   */
+  private void handleAborted(AbortedException aborted) {
+    if (transactionRetryAttempts >= MAX_INTERNAL_RETRIES) {
+      // If the same statement in transaction keeps aborting, then we need to abort here.
+      throwAbortWithRetryAttemptsExceeded();
+    } else if (retryAbortsInternally) {
+      logger.fine(toString() + ": Starting internal transaction retry");
+      while (true) {
+        // First back off and then restart the transaction.
+        try {
+          Thread.sleep(aborted.getRetryDelayInMillis() / 1000);
+        } catch (InterruptedException ie) {
+          Thread.currentThread().interrupt();
+          throw SpannerExceptionFactory.newSpannerException(
+              ErrorCode.CANCELLED, "The statement was cancelled");
+        }
+        try {
+          try {
+            txContext = txManager.resetForRetry();
+          } catch (IllegalStateException e) {
+            if (txManager.getState()
+                == com.google.cloud.spanner.TransactionManager.TransactionState.STARTED) {
+              // In case the transaction manager didn't pick up the abort correctly, we'll rollback
+              // manually and do another try. This seems to be the case if the transaction was
+              // aborted and txContext.executeUpdate(Statement) is called.
+              // TODO: Check whether TransactionContextImpl.executeUpdate(Statement) handles
+              // AbortedException correctly.
+              txManager.rollback();
+              txManager = dbClient.transactionManager();
+              txContext = txManager.begin();
+            }
+          }
+          // Inform listeners about the transaction retry that is about to start.
+          invokeTransactionRetryListenersOnStart();
+          // Then retry all transaction statements.
+          transactionRetryAttempts++;
+          for (RetriableStatement statement : statements) {
+            statement.retry(aborted);
+          }
+          successfulRetries++;
```

Comment:
should this be in the loop, or is it per batch of statements?

Reply:
It is per successful retry of an entire transaction, i.e. per batch of statements, so it should be outside of the loop.